### ISSUE TITLE
cannot find implicit ClassTag after macro expansion

### ISSUE THREAD

**Issue by dsilvasc** [scala/bug#11426]
Given the following program using the `lift` macro from [monadless](http://monadless.io/):

```scala
object Repro {
  val i = lift { 1 }
  val j = lift {
    val v = unlift(i)
    Array("a", "b")
    v + 1
  }
}
```

The `lift` macro fails with `scala.reflect.macros.TypecheckException: not found: value ClassTag` when it tries to `c.typecheck(tree)` in [the macro implementation](https://github.com/monadless/monadless/blob/4ee523c3fb4c87aab76c6f7dc9aead7256f1ca3a/monadless-core/src/main/scala/io/monadless/impl/Macro.scala#L16).

The array companion method being called is `def apply[T: ClassTag](xs: T*): Array[T]`, and according to the exception error message the array line expands to

```scala
scala.Array.apply[String]("a", "b")((ClassTag.apply[String](classOf[java.lang.String]): scala.reflect.ClassTag[String]))
```

Adding `import scala.reflect.ClassTag` to the top of the program fixes type-checking.

Why is the `ClassTag` value reference not the fully-qualified `scala.reflect.ClassTag`?

This was previously reported in the monadless issue tracker as https://github.com/monadless/monadless/issues/6 but it doesn't seem like a bug in the macro.

**Comment by hrhino** [comment]
The macro [uses the much-maligned `resetAllAttrs`](https://github.com/monadless/monadless/blob/4ee523c3fb4c87aab76c6f7dc9aead7256f1ca3a/monadless-core/src/main/scala/io/monadless/impl/Transformer.scala#L398-L403)...

The [tree](https://github.com/scala/scala/blob/v2.12.8/src/compiler/scala/reflect/reify/Taggers.scala#L43-L50) which scalac makes when it materializes a `ClassTag` is
```scala
  def materializeClassTag(tpe: Type): Tree = {
    val tagModule = ClassTagModule
    materializeTag(EmptyTree, tpe, tagModule, {
      val erasure = c.reifyRuntimeClass(tpe, concrete = true)
      val factory = TypeApply(Select(Ident(tagModule), nme.apply), List(TypeTree(tpe)))
      Apply(factory, List(erasure))
    })
  }
```
so the `Ident` pointing to the `ClassTag` object comes pre-typed. By calling `resetAllAttrs`, it's likely that the macro strips that symbol, leaving the innermost bit of that tree as `Ident(TermName("ClassTag"))`, which later needs to be retypechecked and `ClassTag` needs to be resolved.

#5464 is the canonical issue for [the brutality of `resetAllAttrs`](https://github.com/scala/scala/blob/v2.12.8/src/compiler/scala/tools/nsc/ast/Trees.scala#L181-L205), I believe.

tl;dr I'm pretty sure this isn't a scala bug.
